tablecodec/tablecodec.go

```diff
@@ -1300,21 +1300,23 @@ func TruncateIndexValue(v *types.Datum, idxCol *model.IndexColumn, tblCol *model
 	if notStringType {
 		return
 	}
-	originalKind := v.Kind()
-	isUTF8Charset := tblCol.Charset == charset.CharsetUTF8 || tblCol.Charset == charset.CharsetUTF8MB4
-	if isUTF8Charset && utf8.RuneCount(v.GetBytes()) > idxCol.Length {
-		rs := bytes.Runes(v.GetBytes())
+	colValue := v.GetBytes()
+	if tblCol.Charset == charset.CharsetBin || tblCol.Charset == charset.CharsetASCII {
+		// Count character length by bytes if charset is binary or ascii.
+		if len(colValue) > idxCol.Length {
+			// truncate value and limit its length
+			if v.Kind() == types.KindBytes {
+				v.SetBytes(colValue[:idxCol.Length])
+			} else {
+				v.SetString(v.GetString()[:idxCol.Length], tblCol.Collate)
+			}
+		}
+	} else if utf8.RuneCount(colValue) > idxCol.Length {
+		// Count character length by characters for other charset, they all encoding as utf8.
```

Comment:
```suggestion
		// Count character length by characters for other rune-based charsets, they are all internally encoded as UTF-8.
```
I would like to emphasis 'internally encoding' here, and "`UTF-8`" is a better term so it can be distingushed with the `utf8` charset of MySQL/TiDB system.